### PR TITLE
EXP-2476 EmailToPdfConverter: Implement support for embedded images, and fix ampersand handling

### PR DESCRIPTION
**Pull request by alexander-schmidt**
- Embedded CID images are now displayed in the PDF instead of crashing the converter.
- The converter will no longer crash due to non-escaped ampersand characters in the intermediate XML / XHTML.

Note: Those fixes are performed in a sole PR because we only have a single `.eml` file (i.e. the added resource) which triggered _both_ issues.

image

### CODE REVIEW

File: platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java [hunk 1]
```diff
@@ -147,6 +234,61 @@ private static String convertHtmltoXhtml(String html) {
...
+	}
+
+	/**
+	 * Escapes characters that are invalid in HTML.
+	 *
+	 * @param htmlText
+	 *            the original HTML (never <i>null</i>)
+	 * @return the escaped HTML (never <i>null</i>)
+	 */
+	private static String escapeHtml(String htmlText) {
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java:248]:
```suggestion
	private static String escapePlainTextForHtml(String htmlText) {
```

**Review by oliver-richers — CHANGES_REQUESTED** [review]

File: platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java [hunk 2]
```diff
@@ -3,6 +3,7 @@
 import com.softicar.platform.common.io.mime.MimeType;
 import com.softicar.platform.core.module.email.part.chooser.EmailAlternativePartsByTypeChooser;
 import com.softicar.platform.core.module.email.part.sequencer.EmailPartsSequencer;
+import com.sun.mail.util.BASE64DecoderStream;
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java:6]: Now using `InputStream` (which is a super type of `BASE64DecoderStream`).

File: platform-core-module/src/test/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverterTestFiles.java [hunk 3]
```diff
@@ -12,6 +12,8 @@ public interface EmailToPdfConverterTestFiles {
 
 	IResourceSupplier EML_HTML = FACTORY.create("eml-html.eml");
 	IResourceSupplier EML_PLAIN = FACTORY.create("eml-plain.eml");
+	IResourceSupplier EML_WITH_NON_ESCAPED_AMPERSAND_IN_SCRIPT_AND_BASE64_IMAGES =
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/test/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverterTestFiles.java:15]: ok

File: platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java [hunk 4]
```diff
@@ -101,39 +129,98 @@ private boolean isInline(Part part) {
...
 		renderer.setDocumentFromString(xhtml);
 		renderer.layout();
 		renderer.createPDF(output, false);
 	}
 
-	private ByteArrayInputStream convertMsgToEml(Supplier<InputStream> input) {
+	private void assertEmlFormat(MimeMessage message) {
+
+		try {
+			if (message.getHeader("MIME-Version") == null) {
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java:169]: improved the message

File: platform-core-module/src/test/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverterTest.java [hunk 5]
```diff
@@ -101,4 +125,21 @@ private String extractText(byte[] pdfBytes) {
 			throw new RuntimeException(exception);
 		}
 	}
+
+	private long countPixelsWithColor(BufferedImage image, Color expectedColor) {
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/test/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverterTest.java:129]: ...well, not quite. Extracted two methods to `com.softicar.platform.common.ui.image.Images` nevertheless, and added tests.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java [hunk 6]
```diff
@@ -101,39 +129,98 @@ private boolean isInline(Part part) {
...
+				String contentString = (String) content;
+				if (part.isMimeType(MimeType.TEXT_HTML.getIdentifier())) {
+					html.append(contentString);
+				} else if (part.isMimeType(MimeType.TEXT_PLAIN.getIdentifier())) {
+					html.append("<html><body><pre>" + escapeHtml(contentString) + "</pre></body></html>");
+				}
+			}
+
+			if (mimeType.startsWith("image/")) {
+				String base64Data = getBase64Data((BASE64DecoderStream) content);
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java]: yes we can

File: platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java [hunk 7]
```diff
@@ -101,39 +129,98 @@ private boolean isInline(Part part) {
...
+	 *
+	 * @param imageDecoderStream
+	 *            the input stream (never <i>null</i>)
+	 * @return the base64 encoded content of the input stream (never
+	 *         <i>null</i>)
+	 */
+	private static String getBase64Data(BASE64DecoderStream imageDecoderStream) throws IOException {
+
+		byte[] contentBytes = IOUtils.toByteArray(imageDecoderStream);
+		return Base64.getEncoder().encodeToString(contentBytes);
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java:220]: Indeed. Found no such way.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java [hunk 8]
```diff
@@ -101,39 +129,98 @@ private boolean isInline(Part part) {
...
-			String content = (String) part.getContent();
-			if (part.isMimeType(MimeType.TEXT_HTML.getIdentifier())) {
-				renderHtml(content, output);
-			} else if (part.isMimeType(MimeType.TEXT_PLAIN.getIdentifier())) {
-				var html = "<html><body><pre>" + escape(content) + "</pre></body></html>";
-				renderHtml(html, output);
+			String mimeType = determineMimeType(part);
+			Object content = part.getContent();
+
+			if (mimeType.startsWith("text/")) {
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java:138]: ok

File: platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java [hunk 9]
```diff
@@ -67,17 +76,36 @@ public byte[] convertMsgToPdf(Supplier<InputStream> input) {
...
+	private void finishRendering(OutputStream output) {
+
+		String htmlString = html.toString();
+
+		for (Entry<String, Base64Image> entry: base64ImageMap.entrySet()) {
+			String filename = entry.getKey();
+			Base64Image image = entry.getValue();
+			String inlineImage = "data:%s;base64,%s".formatted(image.getContentType(), image.getBase64Data());
+			String search = "src=\"cid:%s\"".formatted(filename);
+			String replacement = "src=\"%s\"".formatted(inlineImage);
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java:103]: yea

File: platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java [hunk 10]
```diff
@@ -101,39 +129,98 @@ private boolean isInline(Part part) {
...
+	 * without decoding the stream. Hence we need to encode it again.
+	 *
+	 * @param imageDecoderStream
+	 *            the input stream (never <i>null</i>)
+	 * @return the base64 encoded content of the input stream (never
+	 *         <i>null</i>)
+	 */
+	private static String getBase64Data(BASE64DecoderStream imageDecoderStream) throws IOException {
+
+		byte[] contentBytes = IOUtils.toByteArray(imageDecoderStream);
```
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java]: every dependency counts towards a thridparty-free world :D

File: platform-common/src/main/java/com/softicar/platform/common/ui/image/Images.java [hunk 11]
```diff
@@ -66,4 +68,64 @@ public static List<BufferedImage> readImages(Supplier<InputStream> inputStreamFa
...
+
+		Objects.requireNonNull(image);
+		Objects.requireNonNull(expectedColor);
+
+		int height = image.getHeight();
+		int width = image.getWidth();
+
+		for (int x = 0; x < width; x++) {
+			for (int y = 0; y < height; y++) {
+				if (!new Color(image.getRGB(x, y)).equals(expectedColor)) {
```
Comment (oliver-richers) [inline platform-common/src/main/java/com/softicar/platform/common/ui/image/Images.java:93]: could be premature optimization but `if(image.getRGB(x, y) != expectedColor.getRGB())` might be more efficient

File: platform-common/src/main/java/com/softicar/platform/common/ui/image/Images.java [hunk 12]
```diff
@@ -66,4 +68,64 @@ public static List<BufferedImage> readImages(Supplier<InputStream> inputStreamFa
...
+		Objects.requireNonNull(image);
+		Objects.requireNonNull(expectedColor);
+
+		int height = image.getHeight();
+		int width = image.getWidth();
+
+		long matchingPixels = 0;
+		for (int x = 0; x < width; x++) {
+			for (int y = 0; y < height; y++) {
+				if (new Color(image.getRGB(x, y)).equals(expectedColor)) {
```
Comment (oliver-richers) [inline platform-common/src/main/java/com/softicar/platform/common/ui/image/Images.java:123]: same possible optimization as above

File: platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java [hunk 13]
```diff
@@ -102,39 +129,98 @@ private boolean isInline(Part part) {
...
+	 *
+	 * @param part
+	 *            the {@link Part} for which the mime type shall be determined
+	 *            (never <i>null</i>)
+	 * @return the mime type in {@code "foo/bar"} format (never <i>null</i>)
+	 */
+	private static String determineMimeType(Part part) throws MessagingException {
+
+		return part.getContentType().replaceAll("\\s+", " ").replaceAll("[; ].*", "");
+	}
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java:204]: I wonder if we should extract this into some `MimeTypeParser`...maybe in a follow-up PR?

File: platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java [hunk 14]
```diff
@@ -102,39 +129,98 @@ private boolean isInline(Part part) {
...
+				if (part.isMimeType(MimeType.TEXT_HTML.getIdentifier())) {
+					html.append(contentString);
+				} else if (part.isMimeType(MimeType.TEXT_PLAIN.getIdentifier())) {
+					html.append("<html><body><pre>" + escapePlainTextForHtml(contentString) + "</pre></body></html>");
+				}
+			}
+
+			else if (part.isMimeType("image/*")) {
+				String base64Data = getBase64Data((InputStream) content);
+				String mimeType = determineMimeType(part);
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java:148]: hmm, for what reason do we bother with the mime type, couldn't we just take it as it is?

**Review by oliver-richers — DISMISSED** [review]

File: platform-common/src/main/java/com/softicar/platform/common/ui/image/Images.java (hunk 11 above)
Comment (alexander-schmidt) [reply platform-common/src/main/java/com/softicar/platform/common/ui/image/Images.java:93]: Yea... probably. Changed accordingly.

File: platform-common/src/main/java/com/softicar/platform/common/ui/image/Images.java (hunk 12 above)
Comment (alexander-schmidt) [reply platform-common/src/main/java/com/softicar/platform/common/ui/image/Images.java:123]: same solution

File: platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java (hunk 13 above)
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java:204]: > Maybe we don't need this method at all?

Yes we do: Unfortunately, `jakarta.mail.Part` only provides `boolean isMimeType(String)` while _not_ providing something like `String getMimeType()`. We can only rely on `String getContentType()` which may or may not contain additional tokens after the sub-type.

See the Javadoc above this method, `determineMimeType`:
image

Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java:204]: We could certainly extract this as soon as there's a (re-)use case.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java (hunk 14 above)
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java:148]: Unfortunately, we cannot:
```
data:%s;base64,%s
     ^
   that String is inserted here
```
...and we can't allow those additional attributes (e.g. `; name="filename.ext"`) from the part content type there.

Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java:148]: ok, got it

**Review by oliver-richers — APPROVED** [review]

**Pull request merged**